static/_posts/2024-04-01-vgaterm-intro-001.md

```diff
@@ -0,0 +1,152 @@
+There are five main components to Vgaterm:
+
+* VGA Timing
+* CPU Pixel Output
+* Displaying a Pixel Byte
+* Digital to Analog Converters (DAC)
+* IO control and video buffer updates
+
+{% include figure.html image="/assets/images/vgaterm-logical-diagram.jpg" caption="Vgaterm logical diagram. Details were omitted to show the high level units" width=800 %}
+
+Repositories containing all our code and schematics:
+* esp32c3 code: [https://github.com/rustbox/esp32c3-vgaterm](https://github.com/rustbox/esp32c3-vgaterm)
+* VGA timing generation code and Programmable Logic Device code: [https://github.com/rustbox/cuddly-robot](https://github.com/rustbox/cuddly-robot)
+* PCB schematics and CAD: [https://github.com/rustbox/vgaterm-schematics](https://github.com/rustbox/vgaterm-schematics)
+
+### VGA Timing
+
+VGA comes in several modes, and we target the 640x480, 60Hz frame rate mode - the base standard. VGA has 3 analog signals between 0 and 0.7 volts representing the amount of Red Green and Blue per pixel. There are also two timing signals `H_SYNC` and `V_SYNC` that tell the monitor what resolution mode to be in and when to sample the analog color signals to actually dispaly a pixel. When the monitor is actively sampling colors and displaying pixels we say that this portion of the frame is `visible`. There is time during a complete frame when this isn't the case and this is _blanking area_. Lastly, the _pixel clock_ speed is the rate at which pixels are shown, and in our case that's 25.175 MHz. In the diagram above, I've labeled a signal `CLK_VIS`, and this is the _pixel clock_ signal during the `visible` time. To see detailed VGA timing information, see [tiny vga](http://tinyvga.com/vga-timing/640x480@60Hz).
+
+:----:|:-----:
+![](/assets/images/vgaterm-timing-front.jpg) | ![](/assets/images/vgaterm-timing-back.jpg)
+<span style="color: {{ site.caption_color }};">Vgaterm Timing circuit top</span> | <span style="color: {{ site.caption_color }};">Vgaterm Timing cicuit bottom</span>
+
+VGA timing is controlled by the "timing board" circuit. This establishes the `H_SYNC` and `V_SYNC` signals that VGA requires, the `START` signal which alerts the CPU to begin  emitting pixels to the FIFO, and the `visible` flag, which is used to compute `CLK_VIS` to read pixel bytes out of the FIFO.
+
+### Pixels Out
+Every frame (60 times a second) the esp32c3 microcontroller needs to output 256,000 bytes (640x400 pixels, 1 byte per pixel) to hardware so they may be displayed. This is the most important software role as any jitter in timing can cause display artifacts.
+
+To cross the clock domains, from the CPU SPI to VGA pixel clock, we used a FIFO memory chip. This lets us write bytes from the CPU while VGA hardware reads bytes exactly when it needs them.
+
+This was clever, but there are a few constraints we had to consider when designing a circuit around the FIFO:
+* The FIFO read and write setup and hold times had to be obeyed by the SPI clock and `CLK_VIS` signals
```

Comment:
```suggestion
*The data going into the FIFO had to obey the setup and hold times within each SPI clock cyle
```

Reply:
can you say a little more for the motivation for the suggestion here? Both the read and the write side have to obey their respective setup and hold times

Reply:
Yeah, there was a stylistic and a technical change compacted into that one suggestion. Splitting them apart:

1. (stylistic) The sentence had been written with a voice that was passive, and when I did that in high school I got angry red pen marks from a fierce little italian man and now I reflexively rewrite passive voice sentences to active. (Actually, did you have the same English teacher as Heather? She does the same thing all the time too)

2. (technical) It feels weird to say that the read/write setup & hold times are a property of the clocks, rather than the data relative to the clock. There is a constraint there, sure, but we added the setup + hold times together and selected a FIFO that was capable of being driven by a 20MHz write / 21.175MHz read signal. I think we even did that on the first try? The much more interesting part (that we struggled with) was how to mesh the data, clock, and up/downstream circuits. 

Maybe you could say a little more about what you want to set up here, and what you hope to leave for the "timing analysis, a journey and erroneous comedy" subsection of whatever in-depth post?

Reply:
Looks like this comment chain got marked as outdated, but the bullet point's wording didn't change.